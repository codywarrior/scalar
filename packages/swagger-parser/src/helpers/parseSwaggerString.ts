import { loadWasmFile } from './loadWasmFile'

/**
 * Parse a Swagger string into a SwaggerObject
 */
export const parseSwaggerString = async (value: any): Promise<any> => {
  return new Promise(async (resolve, reject) => {
    // Load the WASM file.
    await loadWasmFile()

    // Make sure the input is a string.
    const input = typeof value === 'string' ? value : JSON.stringify(value)

    // Parse the input.
    // @ts-ignore
    const result = formatJSON(input)

    // Return the parsed result.
    if (typeof result === 'string') {
      resolve(JSON.parse(result))
      return
    }

    // Throw an error if the result is an error.
    if (result.error) {
      reject(result.error.trim())
      return
    }

    resolve(null)
    return
  })
}
