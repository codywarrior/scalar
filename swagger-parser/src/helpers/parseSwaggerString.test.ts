import { describe, expect, it } from 'vitest'

import exampleInput from './fixtures/petstore.json'
import exampleYaml from './fixtures/swagger.yaml'
import { parseSwaggerString } from './parseSwaggerString'

describe('parseSwaggerString', () => {
  it('parses JSON objects', async () => {
    const result = await parseSwaggerString(exampleInput)

    expect(result).not.toBeNull()
    expect(result).toMatchSnapshot()
  })

  it('parses JSON objects and returns a promise', () => {
    parseSwaggerString(exampleInput).then((result) => {
      expect(result).not.toBeNull()
      expect(result).toMatchSnapshot()
    })
  })

  it('executes the catch callback on error', async () => {
    await parseSwaggerString('FOOBAR').catch((error) => {
      expect(error).not.toBeNull()
      expect(error).toBe(
        'unable to parse JSON. Error spec type not supported by libopenapi, sorry occurred',
      )
    })
  })

  it('parses JSON strings', async () => {
    const result = await parseSwaggerString(JSON.stringify(exampleInput))

    expect(result).not.toBeNull()
    expect(result).toMatchSnapshot()
  })

  it.skip('parses Yaml strings', async () => {
    const result = await parseSwaggerString(exampleYaml)

    expect(result).not.toBeNull()
    expect(result).toMatchSnapshot()
  })

  it('throws an error', async () => {
    expect(async () => {
      await parseSwaggerString('FOOBAR')
    }).rejects.toThrowError()
  })
})
