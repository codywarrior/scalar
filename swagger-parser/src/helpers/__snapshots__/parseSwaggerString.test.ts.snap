// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseSwaggerString > parses JSON objects 1`] = `
{
  "externalDocs": null,
  "info": {
    "license": {
      "name": "MIT",
    },
    "title": "Swagger Petstore",
    "version": "1.0.0",
  },
  "operations": {},
  "servers": [
    {
      "url": "http://petstore.swagger.io/v1",
    },
  ],
  "tags": [
    {
      "description": "",
      "name": "pets",
      "operations": [
        {
          "description": "",
          "httpVerb": "get",
          "information": {
            "operationId": "listPets",
            "parameters": [
              {
                "description": "How many items to return at one time (max 100)",
                "in": "query",
                "name": "limit",
                "schema": {
                  "format": "int32",
                  "type": "integer",
                },
              },
            ],
            "responses": {
              "200": {
                "content": {
                  "application/json": {
                    "schema": {
                      "items": {
                        "properties": {
                          "id": {
                            "format": "int64",
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "tag": {
                            "type": "string",
                          },
                        },
                        "required": [
                          "id",
                          "name",
                        ],
                      },
                      "type": "array",
                    },
                  },
                },
                "description": "An paged array of pets",
                "headers": {
                  "x-next": {
                    "description": "A link to the next page of responses",
                    "schema": {
                      "type": "string",
                    },
                  },
                },
              },
              "default": {
                "content": {
                  "application/json": {
                    "schema": {
                      "properties": {
                        "code": {
                          "format": "int32",
                          "type": "integer",
                        },
                        "message": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "code",
                        "message",
                      ],
                    },
                  },
                },
                "description": "unexpected error",
              },
            },
            "summary": "List all pets",
            "tags": [
              "pets",
            ],
          },
          "name": "List all pets",
          "operationId": "listPets",
          "path": "/pets",
        },
        {
          "description": "",
          "httpVerb": "post",
          "information": {
            "operationId": "createPets",
            "responses": {
              "201": {
                "description": "Null response",
              },
              "default": {
                "content": {
                  "application/json": {
                    "schema": {
                      "properties": {
                        "code": {
                          "format": "int32",
                          "type": "integer",
                        },
                        "message": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "code",
                        "message",
                      ],
                    },
                  },
                },
                "description": "unexpected error",
              },
            },
            "summary": "Create a pet",
            "tags": [
              "pets",
            ],
          },
          "name": "Create a pet",
          "operationId": "createPets",
          "path": "/pets",
        },
        {
          "description": "",
          "httpVerb": "get",
          "information": {
            "operationId": "showPetById",
            "parameters": [
              {
                "description": "The id of the pet to retrieve",
                "in": "path",
                "name": "petId",
                "required": true,
                "schema": {
                  "type": "string",
                },
              },
            ],
            "responses": {
              "200": {
                "content": {
                  "application/json": {
                    "schema": {
                      "items": {
                        "properties": {
                          "id": {
                            "format": "int64",
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "tag": {
                            "type": "string",
                          },
                        },
                        "required": [
                          "id",
                          "name",
                        ],
                      },
                      "type": "array",
                    },
                  },
                },
                "description": "Expected response to a valid request",
              },
              "default": {
                "content": {
                  "application/json": {
                    "schema": {
                      "properties": {
                        "code": {
                          "format": "int32",
                          "type": "integer",
                        },
                        "message": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "code",
                        "message",
                      ],
                    },
                  },
                },
                "description": "unexpected error",
              },
            },
            "summary": "Info for a specific pet",
            "tags": [
              "pets",
            ],
          },
          "name": "Info for a specific pet",
          "operationId": "showPetById",
          "path": "/pets/{petId}",
        },
      ],
    },
  ],
}
`;

exports[`parseSwaggerString > parses JSON objects and returns a promise 1`] = `
{
  "externalDocs": null,
  "info": {
    "license": {
      "name": "MIT",
    },
    "title": "Swagger Petstore",
    "version": "1.0.0",
  },
  "operations": {},
  "servers": [
    {
      "url": "http://petstore.swagger.io/v1",
    },
  ],
  "tags": [
    {
      "description": "",
      "name": "pets",
      "operations": [
        {
          "description": "",
          "httpVerb": "get",
          "information": {
            "operationId": "listPets",
            "parameters": [
              {
                "description": "How many items to return at one time (max 100)",
                "in": "query",
                "name": "limit",
                "schema": {
                  "format": "int32",
                  "type": "integer",
                },
              },
            ],
            "responses": {
              "200": {
                "content": {
                  "application/json": {
                    "schema": {
                      "items": {
                        "properties": {
                          "id": {
                            "format": "int64",
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "tag": {
                            "type": "string",
                          },
                        },
                        "required": [
                          "id",
                          "name",
                        ],
                      },
                      "type": "array",
                    },
                  },
                },
                "description": "An paged array of pets",
                "headers": {
                  "x-next": {
                    "description": "A link to the next page of responses",
                    "schema": {
                      "type": "string",
                    },
                  },
                },
              },
              "default": {
                "content": {
                  "application/json": {
                    "schema": {
                      "properties": {
                        "code": {
                          "format": "int32",
                          "type": "integer",
                        },
                        "message": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "code",
                        "message",
                      ],
                    },
                  },
                },
                "description": "unexpected error",
              },
            },
            "summary": "List all pets",
            "tags": [
              "pets",
            ],
          },
          "name": "List all pets",
          "operationId": "listPets",
          "path": "/pets",
        },
        {
          "description": "",
          "httpVerb": "post",
          "information": {
            "operationId": "createPets",
            "responses": {
              "201": {
                "description": "Null response",
              },
              "default": {
                "content": {
                  "application/json": {
                    "schema": {
                      "properties": {
                        "code": {
                          "format": "int32",
                          "type": "integer",
                        },
                        "message": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "code",
                        "message",
                      ],
                    },
                  },
                },
                "description": "unexpected error",
              },
            },
            "summary": "Create a pet",
            "tags": [
              "pets",
            ],
          },
          "name": "Create a pet",
          "operationId": "createPets",
          "path": "/pets",
        },
        {
          "description": "",
          "httpVerb": "get",
          "information": {
            "operationId": "showPetById",
            "parameters": [
              {
                "description": "The id of the pet to retrieve",
                "in": "path",
                "name": "petId",
                "required": true,
                "schema": {
                  "type": "string",
                },
              },
            ],
            "responses": {
              "200": {
                "content": {
                  "application/json": {
                    "schema": {
                      "items": {
                        "properties": {
                          "id": {
                            "format": "int64",
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "tag": {
                            "type": "string",
                          },
                        },
                        "required": [
                          "id",
                          "name",
                        ],
                      },
                      "type": "array",
                    },
                  },
                },
                "description": "Expected response to a valid request",
              },
              "default": {
                "content": {
                  "application/json": {
                    "schema": {
                      "properties": {
                        "code": {
                          "format": "int32",
                          "type": "integer",
                        },
                        "message": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "code",
                        "message",
                      ],
                    },
                  },
                },
                "description": "unexpected error",
              },
            },
            "summary": "Info for a specific pet",
            "tags": [
              "pets",
            ],
          },
          "name": "Info for a specific pet",
          "operationId": "showPetById",
          "path": "/pets/{petId}",
        },
      ],
    },
  ],
}
`;

exports[`parseSwaggerString > parses JSON strings 1`] = `
{
  "externalDocs": null,
  "info": {
    "license": {
      "name": "MIT",
    },
    "title": "Swagger Petstore",
    "version": "1.0.0",
  },
  "operations": {},
  "servers": [
    {
      "url": "http://petstore.swagger.io/v1",
    },
  ],
  "tags": [
    {
      "description": "",
      "name": "pets",
      "operations": [
        {
          "description": "",
          "httpVerb": "get",
          "information": {
            "operationId": "listPets",
            "parameters": [
              {
                "description": "How many items to return at one time (max 100)",
                "in": "query",
                "name": "limit",
                "schema": {
                  "format": "int32",
                  "type": "integer",
                },
              },
            ],
            "responses": {
              "200": {
                "content": {
                  "application/json": {
                    "schema": {
                      "items": {
                        "properties": {
                          "id": {
                            "format": "int64",
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "tag": {
                            "type": "string",
                          },
                        },
                        "required": [
                          "id",
                          "name",
                        ],
                      },
                      "type": "array",
                    },
                  },
                },
                "description": "An paged array of pets",
                "headers": {
                  "x-next": {
                    "description": "A link to the next page of responses",
                    "schema": {
                      "type": "string",
                    },
                  },
                },
              },
              "default": {
                "content": {
                  "application/json": {
                    "schema": {
                      "properties": {
                        "code": {
                          "format": "int32",
                          "type": "integer",
                        },
                        "message": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "code",
                        "message",
                      ],
                    },
                  },
                },
                "description": "unexpected error",
              },
            },
            "summary": "List all pets",
            "tags": [
              "pets",
            ],
          },
          "name": "List all pets",
          "operationId": "listPets",
          "path": "/pets",
        },
        {
          "description": "",
          "httpVerb": "post",
          "information": {
            "operationId": "createPets",
            "responses": {
              "201": {
                "description": "Null response",
              },
              "default": {
                "content": {
                  "application/json": {
                    "schema": {
                      "properties": {
                        "code": {
                          "format": "int32",
                          "type": "integer",
                        },
                        "message": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "code",
                        "message",
                      ],
                    },
                  },
                },
                "description": "unexpected error",
              },
            },
            "summary": "Create a pet",
            "tags": [
              "pets",
            ],
          },
          "name": "Create a pet",
          "operationId": "createPets",
          "path": "/pets",
        },
        {
          "description": "",
          "httpVerb": "get",
          "information": {
            "operationId": "showPetById",
            "parameters": [
              {
                "description": "The id of the pet to retrieve",
                "in": "path",
                "name": "petId",
                "required": true,
                "schema": {
                  "type": "string",
                },
              },
            ],
            "responses": {
              "200": {
                "content": {
                  "application/json": {
                    "schema": {
                      "items": {
                        "properties": {
                          "id": {
                            "format": "int64",
                            "type": "integer",
                          },
                          "name": {
                            "type": "string",
                          },
                          "tag": {
                            "type": "string",
                          },
                        },
                        "required": [
                          "id",
                          "name",
                        ],
                      },
                      "type": "array",
                    },
                  },
                },
                "description": "Expected response to a valid request",
              },
              "default": {
                "content": {
                  "application/json": {
                    "schema": {
                      "properties": {
                        "code": {
                          "format": "int32",
                          "type": "integer",
                        },
                        "message": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "code",
                        "message",
                      ],
                    },
                  },
                },
                "description": "unexpected error",
              },
            },
            "summary": "Info for a specific pet",
            "tags": [
              "pets",
            ],
          },
          "name": "Info for a specific pet",
          "operationId": "showPetById",
          "path": "/pets/{petId}",
        },
      ],
    },
  ],
}
`;
